pkg/builder/endpoints.go

```diff
@@ -0,0 +1,343 @@
+	// If unset, the builder will create a simple set of endpoints for the target resource.
+	loadBalancing *LoadBalancingSpec
+
+	// clusterID is a way to distinguish endpoints created for the same target
+	// with the same hostname for a different cluster (Optional)
+	clusterID string
+}
+
+// NewEndpointsBuilder returns a new endpoints builder
+func NewEndpointsBuilder(target Target, hostname string) *EndpointsBuilder {
+	return &EndpointsBuilder{
+		target:   target,
+		hostname: hostname,
+	}
+}
+
+func (builder *EndpointsBuilder) WithLoadBalancing(clusterID string, loadbalansing *LoadBalancingSpec) *EndpointsBuilder {
+	builder.clusterID = clusterID
+	builder.loadBalancing = loadbalansing
+	return builder
+}
+
+// Build returns a list of endpoints created based on configuration of the builder
+func (builder *EndpointsBuilder) Build() ([]*externaldns.Endpoint, error) {
+	if err := builder.Validate(); err != nil {
+		return nil, err
+	}
+
+	var endpoints []*externaldns.Endpoint
+
+	// no load-balancing provided, inferring simple strategy
+	if builder.loadBalancing == nil {
+		endpoints = builder.getSimpleEndpoints()
+	} else {
+		// load-balancing present, inferring load-balanced strategy
+		endpoints = builder.getLoadBalancedEndpoints()
+	}
+
+	sort.Slice(endpoints, func(i, j int) bool {
+		return getSetID(endpoints[i]) < getSetID(endpoints[j])
+	})
+	return endpoints, nil
+}
+
+func (builder *EndpointsBuilder) Validate() error {
+	if matched, err := regexp.MatchString(HostnameRegex, builder.hostname); !matched {
+		// This only possible if HostnameRegex is modified.
+		// Leave it here as a precaution
+		if err != nil {
+			return fmt.Errorf("error parsing regexp to match hostname: %w", err)
+		}
+		return fmt.Errorf("invalid hostname")
+	}
+
+	if builder.target == nil {
+		return fmt.Errorf("must provide target")
+	}
+
+	if builder.target.GetAddresses() == nil {
+		return fmt.Errorf("must provide addresses")
+	}
+
+	// following only applicable for load-balancing strategy
+	if builder.loadBalancing != nil {
+		// clusterID must not be an empty string
+		if builder.clusterID == "" {
+			return fmt.Errorf("cluster ID is required")
+		}
+
+		// default weight and geo are required
+		if builder.loadBalancing.DefaultWeight < 0 {
+			return fmt.Errorf("invalid default weight")
+		}
+		if builder.loadBalancing.DefaultGeo == "" {
+			return fmt.Errorf("default geocode is required")
+		}
+
+		// validate custom weights if they were provided
+		if builder.loadBalancing.CustomWeights != nil {
+			for _, customWeight := range builder.loadBalancing.CustomWeights {
+				if customWeight.Weight < 0 {
+					return fmt.Errorf("custom weight cannot be negative")
+				}
+				if customWeight.Selector == nil || customWeight.Selector.MatchLabels == nil && len(customWeight.Selector.MatchLabels) == 0 && customWeight.Selector.MatchExpressions == nil {
+					return fmt.Errorf("custom weight must define non-empty selector")
+				}
+			}
+		}
```

Comment:
Could you rephrase, please?
When i was creating the validation first of all I justified it's presence by "it is a public-facing and it could be fed anything, so I better be as safe as possible" and when going to the checks themselves I opted only to check for things that might case runtime errors. Meaning I'm not checking for the validity of arguments (except for the regex on the hostname) but only their presence. In our case CEL on policy CRD will ensure that validation always succeeds.

Reply:
I see. Well, all the check here are not exactly 1:1 with the ones we have on the DNSPolicy CRD. E.g. we demand default geo to be a string of length at least 2 while I'm here just checking for default empty strings. I was trying to make validations in a "you get what you asked for" scenario and only prevent trivially invalid output (something like `.klb.test.example.com` or negative weight). With this code you can have your default geo to be "A" or "CAD" which are invalid for AWS, while Policy will let you have "CAD".